kvs integration with content store

garlick opened the pull request:
This PR Is not complete but I thought it should get some early high level review concurrently with progress refining it.

The [sophia](http://sphia.org/) embedable database engine is imported into `libflux-internal`, then a new module `cas` (content addressable store) is added which stores arbitrary blobs persistently.  The protocol is simple:  send a blob, get back an SHA1; send an SHA1, get back a blob.  blobs and SHA1's are sent using raw message payloads to avoid the overhead of json encoding.  Files are created in the same temporary directory as the ipc sockets and persist for the lifetime of the session only.  This module is then loaded automatically on rank 0 (only).

To test the cas service, two new operations were added to `flux-kvs`:  putobj and getobj.  These behave similar to the git-hash and get-cat-file utilities, respectively:
```
$ echo hello | ./flux kvs putobj
5f273d69af9e02668217f42bec007fe2492f52f8
$ ./flux kvs getobj 5f273d69af9e02668217f42bec007fe2492f52f8
hello
```

The KVS was then modified so that the rank 0 object cache is expired temporally like other ranks, and the CAS is integrated as the "upstream" cache of rank 0.

This should get the memory utilization of the KVS on rank 0 under control, at the expense of some disk usage (which if it is a ramdisk, well...), and increased commit latency.

I'm not sure if I'm using sophia correctly, particularly with regard to error handling and memory management, as there is not much in the way of examples online.  What I have seems to work at least superficially.  I have enabled lz4 compression of objects.  Some sophia spelunking and stress testing is required to understand how it should be used to best advantage.

There is definitely some work needed to clean things up, factor out duplicated code, and adding tests, but what's here seems to work with make check and manual poking.

grondo commented:
> blobs and SHA1's are sent using raw message payloads to avoid the overhead of json encoding.

This would probably be nice for a lot of other use cases. Do you have handy the snippet of RFC and API detailing this usage? (I'm also fine with just peeking at code, but thought I might ask).

Do you think this is ready for a quick test with many jobs? I'd be willing to poke the tires a bit.

Also looks like an unrelated commit snuck in here, a69534e, want to move that one to a new PR. (I'd also suggest instead of silencing that `fprintf` turn it into a `LOG_DEBUG` message)

garlick commented:
Thanks @grondo, just merged your version of the hwloc fix and I'll plan to rebase this PR.

Yes!  It would be awesome if you'd like to try some scalability testing.  I did some stress testing on a small scale and it held up OK so it would be good to find out if there are some other problems lurking there.

Regarding a snippit for usage, right now you could refer to the implementation of getobj and putobj in `flux-kvs.c`, however this is bare metal message handling and needs to be factored out into something reusable (one of the todos I mentioned above).

garlick commented:
Rebased on current master, and factored common code for converting between string and binary sha1 hashes out of modules/kvs and cmd/flux-kvs to libutil.

grondo commented:
I've poked a bit at this branch and so far it seems really solid!

I ran 4000 jobs through a 4 rank `flux start` session and `flux-broker-0` RSS is only about 2.1G.
I did notice the sophia db was quite a bit larger than I expected:
```
$ du -shc $(flux getattr scratch-directory)
16G     /var/tmp/grondo/flux-225343-cxkE4r
16G     total
```

garlick commented:
Thanks for the testing @grondo.  I will try to figure out why the size is so large.  Maybe I'm updating objects unnecessarily in the db (though I thought that would be squashed in the kvs's object cache).

Here's a rebase with the following changes
* added raw payload versions of `flux_rpc()` and `flux_rpc_get()`, and other encode/decode functions
* tidy up some interfaces in the `flux_msg_*` class
* add an api for the CAS load/store operations (based on extending flux_rpc_t)
* use RPC continuations in the KVS to manage asynchronous CAS RPC's

The CAS api looks like this (it is included with libkvs):
```C
/* store */
flux_rpc_t *cas_store (flux_t h, const void *data, int len);
int cas_store_get_hash (flux_rpc_t *r, void *hash, int *len);

/* load */
flux_rpc_t *cas_load (flux_t h, const void *hash, int len);
int cas_load_get_data (flux_rpc_t *r, void *data, int *len);
int cas_load_get_hash (flux_rpc_t *r, void *hash, int *len);
```

trws commented:
This is looking exceptionally promising.

Based on the numbers I was getting in #462, that's actually smaller than I would expect for 4000 jobs @grondo. You're getting a memory size of 2GB and disk size of 16G where I was getting a memory size of 24GB for rank 0 at that many jobs.  Perhaps this is sophia doing some additional compression for us?

BTW: @garlick, the cas operations look pretty great.  Any chance of either having an atomic add, or a more general atomic with a payload op-type so we can expand on that interface for operations that don't need the infinite consensus property of CAS so we can avoid long-latency retries?

garlick commented:
> Any chance of either having an atomic add, or a more general atomic with a payload op-type so we can expand on that interface for operations that don't need the infinite consensus property of CAS so we can avoid long-latency retries?

@trws what do you mean by "infinite consensus property" ?

garlick commented:
One thing I did just run into in testing with sophia is:
```
[1447699487.235048] cas.err[0]: store: sp_setstring value: sophia/environment/se_v.c:74 value is too big (2097152 limit)
[1447699487.237928] kvs.err[0]: cas_store_completion: Bad file descriptor
```

I haven't tracked down whether this 2MB limit is tunable or a function of other tunables, but having some limit is likely unavoidable.

Another thing I ran into in #474 testing was that lz4 compression slowed down gets by a factor of four.

garlick commented:
I've added some (quite large, but tested) changes to the kvs to modernize its use of the flux API.

I had added a recursive directory cleanup hook to the libutil/cleanup stuff for sophia, which works ok when flux is launched and terminated in the usual way, but seems not to work with sideflux.  Any hints on how to fix this @trws?  (thanks)
```
ok 3 test.event.TestEvent.test_t1_1_unsub
Traceback (most recent call last):
  File "./test_commands/test_runner.t", line 53, in <module>
    run_under_dir(result_set, '../test', 'test')
  File "./test_commands/test_runner.t", line 46, in run_under_dir
    run_tests_with_size(result_set, suite, size)
  File "./test_commands/test_runner.t", line 17, in run_tests_with_size
    tests(result_set)
  File "/opt/python-2.7/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/home/garlick/proj/flux-core/src/bindings/python/test_commands/sideflux.py", line 137, in run_beside_flux
    f.destroy()
  File "/home/garlick/proj/flux-core/src/bindings/python/test_commands/sideflux.py", line 103, in destroy
    shutil.rmtree(self.tmpdir)
  File "/opt/python-2.7/lib/python2.7/shutil.py", line 244, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/opt/python-2.7/lib/python2.7/shutil.py", line 244, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/opt/python-2.7/lib/python2.7/shutil.py", line 253, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/opt/python-2.7/lib/python2.7/shutil.py", line 251, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/tmp/flux-sandbox-nyF9CN/cas/cas'
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/python-2.7/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
Process Process-1:
Traceback (most recent call last):
    func(*targs, **kargs)
  File "/opt/python-2.7/lib/python2.7/multiprocessing/util.py", line 269, in _exit_function
  File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 232, in _bootstrap
    p.join()
  File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 119, in join
    self.run()
  File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 88, in run
    self._target(*self._args, **self._kwargs)
```

trws commented:
What all does it remove? Sideflux sets up a tempdir that it manages and 
passes to flux, so it may be that you're removing the directory that 
side flux set up, then side flux is failing trying to remove it again.

On 16 Nov 2015, at 18:35, Jim Garlick wrote:

> I've added some (quite large, but tested) changes to the kvs to 
> modernize its use of the flux API.
>
> I had added a recursive directory cleanup hook to the libutil/cleanup 
> stuff for sophia, which works ok when flux is launched and terminated 
> in the usual way, but seems not to work with sideflux.  Any hints on 
> how to fix this @trws?  (thanks)
> ```
> ok 3 test.event.TestEvent.test_t1_1_unsub
> Traceback (most recent call last):
> File "./test_commands/test_runner.t", line 53, in <module>
>  run_under_dir(result_set, '../test', 'test')
> File "./test_commands/test_runner.t", line 46, in run_under_dir
>  run_tests_with_size(result_set, suite, size)
> File "./test_commands/test_runner.t", line 17, in run_tests_with_size
>  tests(result_set)
> File "/opt/python-2.7/lib/python2.7/contextlib.py", line 24, in 
> __exit__
>  self.gen.next()
> File 
> "/home/garlick/proj/flux-core/src/bindings/python/test_commands/sideflux.py", 
> line 137, in run_beside_flux
>  f.destroy()
> File 
> "/home/garlick/proj/flux-core/src/bindings/python/test_commands/sideflux.py", 
> line 103, in destroy
>  shutil.rmtree(self.tmpdir)
> File "/opt/python-2.7/lib/python2.7/shutil.py", line 244, in rmtree
>  rmtree(fullname, ignore_errors, onerror)
> File "/opt/python-2.7/lib/python2.7/shutil.py", line 244, in rmtree
>  rmtree(fullname, ignore_errors, onerror)
> File "/opt/python-2.7/lib/python2.7/shutil.py", line 253, in rmtree
>  onerror(os.rmdir, path, sys.exc_info())
> File "/opt/python-2.7/lib/python2.7/shutil.py", line 251, in rmtree
>  os.rmdir(path)
> OSError: [Errno 39] Directory not empty: 
> '/tmp/flux-sandbox-nyF9CN/cas/cas'
> ^CError in atexit._run_exitfuncs:
> Traceback (most recent call last):
> File "/opt/python-2.7/lib/python2.7/atexit.py", line 24, in 
> _run_exitfuncs
> Process Process-1:
> Traceback (most recent call last):
>  func(*targs, **kargs)
> File "/opt/python-2.7/lib/python2.7/multiprocessing/util.py", line 
> 269, in _exit_function
> File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 
> 232, in _bootstrap
>  p.join()
> File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 
> 119, in join
>  self.run()
> File "/opt/python-2.7/lib/python2.7/multiprocessing/process.py", line 
> 88, in run
>  self._target(*self._args, **self._kwargs)
> ```
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/471#issuecomment-157220480

garlick commented:
It's removing a cas subdirectory in the tmpdir that contains the sockets.  I extended the cleanup stuff to provide a way to recursively remove that directory since I can't predict what files will be in there.

trws commented:
That shouldn't be causing problems... I'll have to run some tests to track this down.



Sent with Good (www.good.com)

________________________________
From: Jim Garlick
Sent: Monday, November 16, 2015 5:57:37 PM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] implement kvs backing store using sophia (#471)


It's removing a cas subdirectory in the tmpdir that contains the sockets. I extended the cleanup stuff to provide a way to recursively remove that directory since I can't predict what files will be in there.

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-157236342>.

grondo commented:
@garlick, the latest Travis failure:
```
not ok 95 - reactor exited normally with Protocol error
#     Failed test (...ore/flux-core-0.1.0/_build/t/../../t/lua/t1002-kvs.t at line 168)
#          got: nil
#     expected: 0
```

Was something I was also seeing in my development branch, so it isn't specific to your work here. I was trying to run the problem down (an error in Lua bindings most likely), but then the issue went away and I could not reproduce anymore.

I will try issuing a rebuild in travis and see if the problem goes away, and try recreating the problem elsewhere (tomorrow).

Apologies.

trws commented:
> @trws what do you mean by "infinite consensus property" ?

I was referring to the "consensus number" applied to CAS or LL/SC as 
opposed to something like fetch-and-add or other atomic operations.  In 
general FAA or test and set or similar have a consensus number of two, 
meaning they can cause a maximum of two threads to reach consensus on a 
state, where CAS can do so for an infinite number.  This is a very good 
thing when you need it, but it can only do that because of the fact that 
it can, by definition, fail and cost an arbitrary number of retries in 
each caller.

All I meant to get at was, CAS is a very good thing since one can use it 
to implement literally any other atomic construct, but it's frequently 
more expensive than necessary to get the job done.

garlick commented:
I think we're on different pages here @trws.  My use of CAS was "content addressable store".  Maybe you are referring to "compare and set"?

trws commented:
Yup, completely missed the boat there.  Sorry. ><

Perhaps a different acronym/name would reduce confusion?   Especially if we're ever going to have atomic kvs operations?

trws commented:
I got a bit of insight on the sideflux issue. It looks like something in the the content addressable storage directory can't be removed by the python code, so it throws an exception when that fails. I pushed up a fix that should let it pass to a like-named branch on my fork, but I'm not sure how much trash will get left around by letting the python tmpdir cleanup fail. In principle it should be none now, but it would be best to test it.



Sent with Good (www.good.com)

________________________________
From: Mark Grondona
Sent: Tuesday, November 17, 2015 8:48:29 AM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] implement kvs backing store using sophia (#471)


@garlick<https://github.com/garlick>, the latest Travis failure:

not ok 95 - reactor exited normally with Protocol error
#     Failed test (...ore/flux-core-0.1.0/_build/t/../../t/lua/t1002-kvs.t at line 168)
#          got: nil
#     expected: 0


Was something I was also seeing in my development branch, so it isn't specific to your work here. I was trying to run the problem down (an error in Lua bindings most likely), but then the issue went away and I could not reproduce anymore.

I will try issuing a rebuild in travis and see if the problem goes away, and try recreating the problem elsewhere (tomorrow).

Apologies.

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-157427985>.

garlick commented:
I've added sophia and sqlite results to #474 for comparison with the in-memory hashes.

sqlite is looking appealing to me in part because it can be configured to have a very small memory footprint and still function well (not tuning its default cache size, the test with 16M objects only used 2.6MB!)  I haven't looked at the maximum object size.  The hardwired 2mb limit is a major complication for using sophia here.

garlick commented:
The current code uses the backing store synchronously but it would actually be reasonable IMHO to use it asynchronously, e.g. return to the KVS caller before object is on disk, as long as we keep the dirty bit on the cache entry until it is.

trws commented:
The sqlite numbers look good for memory size, but what was the backing 
store there?  Big advantages to sqlite are that it is available almost 
everywhere and it's possible to clear out dead data if and when we can 
do that, where sophia is append-only. Also, I would want to see how the 
inserts are being done before looking too hard at that insert overhead.  
The most straightforward interfaces also tend to be the slowest, and 
there is usually quite a bit of room for optimization in table design 
and query management.

On 18 Nov 2015, at 15:19, Mark Grondona wrote:

> Is it because the store operation would be asynchronous that you're 
> not worried about the 20x performance hit for Insert with sqlite?
>
> It might also help to document for the persistence cases what the 
> backing store is (disk, ssd, tmpfs).
>
> Really neat work here!
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/471#issuecomment-157866492

garlick commented:
I've rebased and pushed the result, with the sqlite test baked into the `hashtest` program so @trws can see how I constructed the tables and set various pragmas and so forth.

I had created the sqlite files in /tmp on my workstation which is an ext4 file system. 

Given that we're currently treating the KVS back end as non-persistent beyond the life of the instance, we can easily swap in different back ends for different needs.  I think this PR ought to deliver sqlite, sophia, and null (memory) back ends and make them configurable and possibly tunable at runtime.

I think also the zhash based cache needs to switch to one of the contenders documented here, and should be operated in "write back" mode with respect to the back end store, possibly with a hard cap on the number of items allowed by be dirty, after which it switches into "write through" mode.  Finally, "cas" should be renamed to something better, maybe "odb" for object database?

grondo commented:
Nice! I didn't have a problem with the name"cas" fwiw. I've seen CAS as the acronym for Content Addressable Storage at least as much as compare-and-swap (which in apis usually seems to be fully spelled out). Also with the fully function names `cas_load` and `cas_store` it seemed pretty obvious.

That being said, I don't mind "odb" either though my fingers will want to type obd, and there already seems to be a lot of references to "object databases" as object persistence for object oriented languages, which might confuse someone from that domain. ;-)

To fully avoid this confusion @dun suggested Jim's Object Emporium, or `joe`

trws commented:
It's probably just divergent awareness, but I had never seen cas as an 
acronym for content addressable store before (hence my now obvious 
mistake, sorry again...).  I find joe pretty amusing, but perhaps 
something descriptive like content_{load,store} if we can buy 4 extra 
characters?  Or failing that, perhaps "content addressable database" 
(cad)? I realize the latter is also a name conflict, but if someone is 
building a computer aided drafting tool with the flux toolkit they may 
have deeper problems... =)

BTW: thanks for putting up the test suite @garlick, I'll take a look at 
it after my presentation at the Green500 BoF is done.

On 18 Nov 2015, at 18:10, Mark Grondona wrote:

> Nice! I didn't have a problem with the name"cas" fwiw. I've seen CAS 
> as the acronym for Content Addressable Storage at least as much as 
> compare-and-swap (which in apis usually seems to be fully spelled 
> out). Also with the fully function names `cas_load` and `cas_store` it 
> seemed pretty obvious.
>
> That being said, I don't mind "odb" either though my fingers will want 
> to type obd, and there already seems to be a lot of references to 
> "object databases" as object persistence for object oriented 
> languages, which might confuse someone from that domain. ;-)
>
> To fully avoid this confusion @dun suggested Jim's Object Emporium, or 
> `joe`
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/471#issuecomment-157905139

grondo commented:
@trws, agreed I mostly kidding about the naming. I meant to say I'd be fine with almost anything.

garlick commented:
I've pushed an update with the following changes
* renamed `cas_load()` and `cas_store()` et al to `content_load()` and `content_store()`
* renamed cas module to `content-sophia`
* added `content-sqlite`
* added MODULE_SERVICE macro and broker routing logic which allows two modules (different names) to provide the same service.  Here both content-sqlite and content-sophia answer to `content.store` and `content.load`.
* sqlite is now the default (and sqlite3-devel is a new prereq)

I haven't really tried the sqlite backend with other than 'make check' , and it's not doing any batching of commits at the moment so probably slow.

Still quite a bit to do here but wanted to checkpoint.

grondo commented:
> added MODULE_SERVICE macro and broker routing logic which allows two modules (different names) to provide the same service. Here both content-sqlite and content-sophia answer to content.store and content.load.

This is nice! Separating service topic string from module name is a great improvement. In this case, how would one replace `content-sqlite` with `content-sophia` -- looks like `svc_add` doesn't allow services to be overidden (which would seem to the right approach in this case), yet if you specify `content-sophia` on cmdline won't it clash with `content-sqlite` and cause broker to `err_exit`?

I realize this stuff is in early stages, but perhaps we'll eventually need `svc_push` `svc_pop` and
a way for services to get a usecount or something so a push/pop is not allowed? Apologies if this
is in the works or out of scope for this PR.

Also, for the new `struct internal_service`, we might want to add a `nodeset_t` member listing ranks on which to restrict the internal service (with perhaps `NULL` for any rank?). If we decide to keep something like `cmb.seq` it will want to target itself for rank 0 only and I can imagine other cases where a subset of ranks may want to load an internal service.

In any event, this specific PR is headed in a nice direction!

garlick commented:
It's a bit awkward but this works
```
./flux start -o,--exclude=content-sqlite,--module=content-sophia
```
Maybe it would be better if the `--exclude` weren't necessary and the conflicting module just "stole" the "content" route?  Hmm.  Yeah, there's definitely a usability issue here.

trws commented:
What I recall us discussing for this was to eventually allow layering of these things, so the last loaded gets all of the the messages, then maybe has a way to do the equivalent of DYLD_NEXT to pass it on and catch the reply for further processing. Even if it could just override for now it would be a great step on that path. This is looking great overall by the way, very nice work @garlick!



Sent with Good (www.good.com)

________________________________
From: Jim Garlick
Sent: Thursday, November 19, 2015 9:09:21 PM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] implement kvs backing store using sophia (#471)


It's a bit awkward but this works

./flux start -o,--exclude=content-sqlite,--module=content-sophia


Maybe it would be better if the --exclude weren't necessary and the conflicting module just "stole" the "content" route? Hmm. Yeah, there's definitely a usability issue here.

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-158281098>.

garlick commented:
Interesting use cases for multiple service providers that are pertinent to this PR might be:
* load/unload a `content-cache` module "in front of" `content-whatever` to seamlessly enable/disable caching.  The cache module would intercept load requests and respond to the ones that are in cache, or forward the ones that need to go to the next content instance.
* load `content-sophia` in front of `content-sqlite` and transfer any data already stored from sqlite to sophia then ignore or even shutdown sqlite (would need a content iterator of some sort then)

Possibly this is getting a little out of scope for this PR though...

grondo commented:
I do like the idea of swapping in/out `content-cache` providers, but that does seem beyond the scope of this PR. However, it does seem like an easy way to force one `content-*` provider over another is a good idea. (Though I'm fine with the explicit `--exclude` `--module` combo that works for now)

garlick commented:
I just pushed a change that implements a stack of service names, with the module at the top of the stack receiving all messages addressed to the service:
```
./flux start -o,--module=content-sophia
```

Both modules will have created their empty database files under tmpdir/content.  Also loss of content will result if content-sophia is loaded after the kvs has already stored data in content-sqlite (e.g. if loaded after session initialization using `flux module load`.  So although it's more flexible, there are more ways to shoot yourself in the foot.

garlick commented:
For the record, PR #478 "raw rpc" was split off this one and merged.

I'm working on a new PR that adds the content service and API (standalone).

This PR will then be reworked to contain KVS cleanup and integration with the content service.  It'll stay parked as is for a bit until that happens.

garlick commented:
Now that the content service PR #493 has been merged, this PR has been rebased on master, and now includes only
* modifications to KVS to avoid deprecated interfaces
* some KVS refactoring
* the hashtest test (somewhat orthogonal to this PR actually...)
* switch KVS over to object cache

I've parked it here, and appears to be passing tests, but it still needs some auditing by me before I ask others to review.

garlick commented:
Further work incorporating RFC 11 changes were parked here, but it's in a bad state at the moment - there's a known bug that I'm trying to run down, hence travis-ci is sad.

garlick commented:
I just rebased this PR to eliminate the treeref changes, as the amount of change here was getting out of hand.  To deal with the max blob size issue, where KVS directories and values were limited to 1m, I changed it to 1g temporarily. I also made it a broker attribute so it only has to be hardwired in one place.

This PR now contains:
* significant, though by no means complete KVS cleanup
* content cache acts as a backing store for KVS
* optionally, `content-sophia` or `content-sqlite` modules can be loaded to offload the content cache to disk
* some additional KVS tests for large values and directories

The bottom line (hopefully) is that we can run many jobs through Flux without running rank 0 out of memory.  That is, if you load `content-sqlite` (recommended) or `content-sophia`.

I'd like to propose merging this as-is, and then continue with the RFC 11 treeref changes in another PR.

I think first though I'd like to try valgrind before and after to see if I introduced any new leaks.  I need to dig up the rune for starting the broker under valgrind (if anyone has it handy).  First braindead attempt didn't yield much info:
```
valgrind --leak-check=yes .libs/lt-flux-broker
```

trws commented:
That all sounds good. You may want to try running it under heap  sanitizer or address sanitizer, I had some really good luck with finding some memory leaks in flux that way, and it stayed reasonably responsive during the tests.


________________________________
From: Jim Garlick
Sent: Thursday, January 14, 2016 12:48:35 PM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] kvs integration with content store (#471)


I just rebased this PR to eliminate the treeref changes, as the amount of change here was getting out of hand. To deal with the max blob size issue, where KVS directories and values were limited to 1m, I changed it to 1g temporarily. I also made it a broker attribute so it only has to be hardwired in one place.

This PR now contains:

  *   significant, though by no means complete KVS cleanup
  *   content cache acts as a backing store for KVS
  *   optionally, content-sophia or content-sqlite modules can be loaded to offload the content cache to disk
  *   some additional KVS tests for large values and directories

The bottom line (hopefully) is that we can run many jobs through Flux without running rank 0 out of memory. That is, if you load content-sqlite (recommended) or content-sophia.

I'd like to propose merging this as-is, and then continue with the RFC 11 treeref changes in another PR.

I think first though I'd like to try valgrind before and after to see if I introduced any new leaks. I need to dig up the rune for starting the broker under valgrind (if anyone has it handy). First braindead attempt didn't yield much info:

valgrind --leak-check=yes .libs/lt-flux-broker


—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-171775033>.

grondo commented:
> valgrind --leak-check=yes .libs/lt-flux-broker

I had been running with some success using `valgrind --trace-children=yes --child-silent-after-fork=yes`.
I was actually running that on `flux-start` and saving output to a log file then going through it. Valgrind options `--trace-children-skip=PATTERN` or `--trace-children-skip-by-arg` might also be helpful.

dongahn commented:
This reminds me, I actually have a patch for Sanitizer support for flux-core that need to submit…

Also, yesterday, I ran TotalView memscape and found a bunch of leaks. I fixed most of it within my code in sched but there are some within core. Like before, there are some false positives but in general those leaks reported by memscape are something to look.

Dong



----------------------------------------------------------
Dong H. Ahn
925-422-1939
Development Environment  Group
Livermore Computing Division
Lawrence Livermore National Laboratory

From: Tom Scogland <notifications@github.com<mailto:notifications@github.com>>
Reply-To: flux-framework/flux-core <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, January 14, 2016 at 1:04 PM
To: flux-framework/flux-core <flux-core@noreply.github.com<mailto:flux-core@noreply.github.com>>
Subject: Re: [flux-core] kvs integration with content store (#471)

That all sounds good. You may want to try running it under heap sanitizer or address sanitizer, I had some really good luck with finding some memory leaks in flux that way, and it stayed reasonably responsive during the tests.


________________________________
From: Jim Garlick
Sent: Thursday, January 14, 2016 12:48:35 PM
To: flux-framework/flux-core
Cc: Scogland, Thomas Richard William
Subject: Re: [flux-core] kvs integration with content store (#471)


I just rebased this PR to eliminate the treeref changes, as the amount of change here was getting out of hand. To deal with the max blob size issue, where KVS directories and values were limited to 1m, I changed it to 1g temporarily. I also made it a broker attribute so it only has to be hardwired in one place.

This PR now contains:

* significant, though by no means complete KVS cleanup
* content cache acts as a backing store for KVS
* optionally, content-sophia or content-sqlite modules can be loaded to offload the content cache to disk
* some additional KVS tests for large values and directories

The bottom line (hopefully) is that we can run many jobs through Flux without running rank 0 out of memory. That is, if you load content-sqlite (recommended) or content-sophia.

I'd like to propose merging this as-is, and then continue with the RFC 11 treeref changes in another PR.

I think first though I'd like to try valgrind before and after to see if I introduced any new leaks. I need to dig up the rune for starting the broker under valgrind (if anyone has it handy). First braindead attempt didn't yield much info:

valgrind --leak-check=yes .libs/lt-flux-broker


—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-171775033>.

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-171779292>.

grondo commented:
> Also, yesterday, I ran TotalView memscape and found a bunch of leaks. I fixed most of it within my code in sched but there are some within core. Like before, there are some false positives but in general those leaks reported by memscape are something to look.

I forget, are these being tracked in an existing issue?

dongahn commented:
No. There was one that @garlick determined to be a false positive. But there are more that I found yesterday. I’m hurriedly preparing for a big sched PR at the moment, but we should discuss this next Monday as to how best I can provide the tooling information to you folks.

Dong


----------------------------------------------------------
Dong H. Ahn
925-422-1939
Development Environment  Group
Livermore Computing Division
Lawrence Livermore National Laboratory

From: Mark Grondona <notifications@github.com<mailto:notifications@github.com>>
Reply-To: flux-framework/flux-core <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Thursday, January 14, 2016 at 1:48 PM
To: flux-framework/flux-core <flux-core@noreply.github.com<mailto:flux-core@noreply.github.com>>
Cc: Dong Ahn <ahn1@llnl.gov<mailto:ahn1@llnl.gov>>
Subject: Re: [flux-core] kvs integration with content store (#471)


Also, yesterday, I ran TotalView memscape and found a bunch of leaks. I fixed most of it within my code in sched but there are some within core. Like before, there are some false positives but in general those leaks reported by memscape are something to look.

I forget, are these being tracked in an existing issue?

—
Reply to this email directly or view it on GitHub<https://github.com/flux-framework/flux-core/pull/471#issuecomment-171791815>.

grondo commented:
For now, opening a Github issue per obvious case or set of cases would be
best I think.


On Thu, Jan 14, 2016 at 1:51 PM, Dong H. Ahn <notifications@github.com>
wrote:

> No. There was one that @garlick determined to be a false positive. But
> there are more that I found yesterday. I’m hurriedly preparing for a big
> sched PR at the moment, but we should discuss this next Monday as to how
> best I can provide the tooling information to you folks.
>
> Dong
>
>
> ----------------------------------------------------------
> Dong H. Ahn
> 925-422-1939
> Development Environment Group
> Livermore Computing Division
> Lawrence Livermore National Laboratory
>
> From: Mark Grondona <notifications@github.com<mailto:
> notifications@github.com>>
> Reply-To: flux-framework/flux-core <reply@reply.github.com<mailto:
> reply@reply.github.com>>
> Date: Thursday, January 14, 2016 at 1:48 PM
> To: flux-framework/flux-core <flux-core@noreply.github.com<mailto:
> flux-core@noreply.github.com>>
> Cc: Dong Ahn <ahn1@llnl.gov<mailto:ahn1@llnl.gov>>
> Subject: Re: [flux-core] kvs integration with content store (#471)
>
>
> Also, yesterday, I ran TotalView memscape and found a bunch of leaks. I
> fixed most of it within my code in sched but there are some within core.
> Like before, there are some false positives but in general those leaks
> reported by memscape are something to look.
>
> I forget, are these being tracked in an existing issue?
>
> —
> Reply to this email directly or view it on GitHub<
> https://github.com/flux-framework/flux-core/pull/471#issuecomment-171791815>.
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/471#issuecomment-171792417>
> .
>

grondo commented:
@garlick: looks great to me!
I'm just building a copy of your branch and kicking the tires a bit

grondo commented:
I ran approx 1000 jobs through a default session, then ran `flux module load content-sophia`.
It appeared everything worked, but I got a massive stream of these errors from the flux logging system:
```
[1452810608.258939] broker.err[0]: content store: Device or resource busy
[1452810608.258943] broker.err[0]: content store: Device or resource busy
[1452810608.258947] broker.err[0]: content store: Device or resource busy
[1452810608.258950] broker.err[0]: content store: Device or resource busy
[1452810608.258954] broker.err[0]: content store: Device or resource busy
[1452810608.258958] broker.err[0]: content store: Device or resource busy
[1452810608.258962] broker.err[0]: content store: Device or resource busy
[1452810608.258966] broker.err[0]: content store: Device or resource busy

```

garlick commented:
I appreciate that.

At the moment I'm stumped on these leaks that appear to have been introduced around 31e78ab7f134cc9da51215f32bb30c7ffd780ec7 (factor out object cache to its own class):
```
==20179== 606 (216 direct, 390 indirect) bytes in 3 blocks are definitely lost in loss record 1,117 of 1,347
==20179==    at 0x4A057BB: calloc (vg_replace_malloc.c:593)
==20179==    by 0x3174C03871: json_object_new (json_object.c:178)
==20179==    by 0x3174C03C66: json_object_new_string_len (json_object.c:647)
==20179==    by 0x3174C04BE3: json_tokener_parse_ex (json_tokener.c:435)
==20179==    by 0x3174C0654D: json_tokener_parse_verbose (json_tokener.c:162)
==20179==    by 0x3174C065AD: json_tokener_parse (json_tokener.c:150)
==20179==    by 0x8C3EBE5: put_request_cb (shortjson.h:293)
==20179==    by 0x4EA58D4: handle_cb (dispatch.c:341)
==20179==    by 0x4EC0E2C: ev_invoke_pending (ev.c:3088)
==20179==    by 0x4EC5459: ev_run (ev.c:3488)
==20179==    by 0x4EA4B20: flux_reactor_run (reactor.c:129)
==20179==    by 0x8C3F17E: mod_main (kvs.c:1792)
==20179==
==20179== 858 (72 direct, 786 indirect) bytes in 1 blocks are definitely lost in loss record 1,165 of 1,347
==20179==    at 0x4A057BB: calloc (vg_replace_malloc.c:593)
==20179==    by 0x3174C03871: json_object_new (json_object.c:178)
==20179==    by 0x3174C0461A: json_object_new_object (json_object.c:354)
==20179==    by 0x8C3D9A1: copydir (kvs.c:481)
==20179==    by 0x8C3DA53: commit_apply_start (kvs.c:568)
==20179==    by 0x8C3F9C3: commit_apply_all (kvs.c:620)
==20179==    by 0x8C403B5: commit_request_cb (kvs.c:1421)
==20179==    by 0x4EA58D4: handle_cb (dispatch.c:341)
==20179==    by 0x4EC0E2C: ev_invoke_pending (ev.c:3088)
==20179==    by 0x4EC5459: ev_run (ev.c:3488)
==20179==    by 0x4EA4B20: flux_reactor_run (reactor.c:129)
==20179==    by 0x8C3F17E: mod_main (kvs.c:1792)
==20179==
==20179== 4,683 (216 direct, 4,467 indirect) bytes in 3 blocks are definitely lost in loss record 1,287 of 1,347
==20179==    at 0x4A057BB: calloc (vg_replace_malloc.c:593)
==20179==    by 0x3174C03871: json_object_new (json_object.c:178)
==20179==    by 0x3174C0461A: json_object_new_object (json_object.c:354)
==20179==    by 0x8C3DCAD: commit_apply_dirents (kvs.c:525)
==20179==    by 0x8C3FA30: commit_apply_all (kvs.c:629)
==20179==    by 0x8C403B5: commit_request_cb (kvs.c:1421)
==20179==    by 0x4EA58D4: handle_cb (dispatch.c:341)
==20179==    by 0x4EC0E2C: ev_invoke_pending (ev.c:3088)
==20179==    by 0x4EC5459: ev_run (ev.c:3488)
==20179==    by 0x4EA4B20: flux_reactor_run (reactor.c:129)
==20179==    by 0x8C3F17E: mod_main (kvs.c:1792)
==20179==    by 0x40CE4D: module_thread (module.c:140)
==20179==
```

garlick commented:
> [1452810608.258939] broker.err[0]: content store: Device or resource busy

Hmm, one way that can happen is if the matchtag pool for the broker handle is out of tags, which could happen if many, many rpcs were either pending or not cleaned up properly.  Having a look....

Does it happen with sqlite?

garlick commented:
Where many, many = 64k.

grondo commented:
> Does it happen with sqlite?

I'll test that out.

grondo commented:
> Does it happen with sqlite?
> > I'll test that out.

Yes, I simply did
```
flux module remove content-sophia
[long pause]
flux module load content-sqlite
```
And saw the same stream of errors.

SteVwonder commented:
For what is worth, I have been running under valgrind like:
```flux valgrind --log-file=./"memcheck.%p.log" --tool=memcheck --leak-check=full --trace-children=no --num-callers=30 --leak-resolution=med .libs/flux-broker $(BROKER-OPTS)```
I think this is very similar to what grondo was doing, but you don't have to trace-children (keeping your working dir clean of extraneous logs).  I have also found that leak-resolution=med does a pretty decent job of merging similar stack traces without going too far.

grondo commented:
I subsequently tried to read wreck job data back for all jobs with `flux wreck ls` and hit hang :crying_cat_face:
```
(flux-239909) grondo@ipa15:~/git/f$ flux wreck ls
    ID NTASKS STATE                    START      RUNTIME    RANKS COMMAND
     1      4 complete   2016-01-14T14:07:38       0.030s    [0-3] hostname
     2      4 complete   2016-01-14T14:07:38       0.024s    [0-3] hostname
     3      4 complete   2016-01-14T14:07:38       0.025s    [0-3] hostname
     4      4 complete   2016-01-14T14:07:38       0.025s    [0-3] hostname
     5      4 complete   2016-01-14T14:07:38       0.025s    [0-3] hostname
     6      4 complete   2016-01-14T14:07:38       0.025s    [0-3] hostname
     7      4 complete   2016-01-14T14:07:38       0.026s    [0-3] hostname
[hang]
```

`flux-broker-0` is utilizing 100% cpu. I attached to get a backtrace, but gdb quickly hung on me. :angry:  However, from what I saw all threads were idle except the main broker thread which was here (when I interrupted it anyway)

```
Thread 1 (Thread 0x2aaaad519940 (LWP 239966)):
#0  0x00002aaaab657b29 in zmsg_last () from /usr/lib64/libczmq.so.1
#1  0x00002aaaaaf66fab in flux_msg_get_type (msg=<value optimized out>, 
    type=0x7fffffffd67c) at message.c:236
#2  0x00002aaaaaf687da in flux_msg_cmp (msg=0x41755c20, match=...)
    at message.c:423
#3  0x00002aaaaaf6484e in dispatch_message (r=0x6348a0, 
    hw=<value optimized out>, revents=<value optimized out>, arg=0x6419f0)
    at dispatch.c:340
#4  handle_cb (r=0x6348a0, hw=<value optimized out>, 
    revents=<value optimized out>, arg=0x6419f0) at dispatch.c:435
#5  0x00002aaaaaf7fdbd in ev_invoke_pending (loop=0x2aaaab1e63a0) at ev.c:3088
#6  0x00002aaaaaf843ea in ev_run (loop=0x2aaaab1e63a0, flags=0) at ev.c:3488
#7  0x00002aaaaaf63ab1 in flux_reactor_run (r=0x6348a0, flags=0)
    at reactor.c:129
#8  0x00000000004095fd in main (argc=<value optimized out>, 
    argv=<value optimized out>) at broker.c:694
```

grondo commented:
> For what is worth, I have been running under valgrind like:
@SteVwonder:  awesome advice thanks!
Do you have a `valgrind` script in FLUX_EXEC_PATH, or are you doing `flux /path/to/valgrind ...`?

Thanks those other options are great. I didn't know about --leak-resolution.

garlick commented:
Just ran 1000 4-proc jobs through on my desktop with the sqlite backend and didn't hit the problem.  Note the zero dirty entries below, which indicates the backing store is working.
Also looks like we probably purged the cache based on the configured max size of 16mb.
The rss of my broker is 322mb according to top.
```
$ flux lsattr -v|grep content
content-acct-dirty                      0
content-acct-entries                    7726
content-acct-size                       16728493
content-acct-valid                      7726
content-backing                         content-sqlite
content-blob-size-limit                 1073741824
content-hash                            sha1
content-purge-large-entry               256
content-purge-old-entry                 5
content-purge-target-entries            1048576
content-purge-target-size               16777216
```
Repeating the test with sophia, I didn't hit the problem.  RSS of broker 0 is 1.9g though!
```
content-acct-dirty                      0
content-acct-entries                    6785
content-acct-size                       16734986
content-acct-valid                      6785
content-backing                         content-sophia
content-blob-size-limit                 1073741824
content-hash                            sha1
content-purge-large-entry               256
content-purge-old-entry                 5
content-purge-target-entries            1048576
content-purge-target-size               16777216
```
What file system type is your /tmp (where sohpia data would go)?  Mine's ext4

garlick commented:
Incidentally, running 1000 4-proc jobs just tipped the number of requests received by sophia past the 64k mark:
```
flux comms-stats  content-sophia
{
  "#request (tx)":68221,
  "#request (rx)":68216,
  "#response (tx)":68215,
  "#response (rx)":6,
  "#event (tx)":0,
  "#event (rx)":0,
  "#keepalive (tx)":0,
  "#keepalive (rx)":0
}
```
So if sophia was stuck from the beginning, then that number of outstanding requests would indeed exhaust the matchtag pool.

grondo commented:
> What file system type is your /tmp (where sohpia data would go)? Mine's ext4

I was on ipa15 -- looks like `/tmp` is ext3.

I just tried test again (run 1000 jobs without content-* provider), then loaded `content-sqlite` instead of sophia. Got the same set of errors. (Actually I may have run 2000+ jobs, oops). Broker was unresponsive for awhile, but seems to have recovered.

grondo commented:
```
$ flux lsattr -v | grep content
content-acct-dirty                      96389
content-acct-entries                    106132
content-acct-size                       1771201893
content-acct-valid                      106132
content-backing                         content-sqlite
content-blob-size-limit                 1073741824
content-hash                            sha1
content-purge-large-entry               256
content-purge-old-entry                 5
content-purge-target-entries            1048576
content-purge-target-size               16777216
```

grondo commented:
This time I also got some kvs errors in the mix, though I assume they're completely related:
```
[1452816351.649394] kvs.err[1]: content_store_completion: Device or resource busy
[1452816351.707019] kvs.err[2]: content_store_completion: Device or resource busy
[1452816351.649617] broker.err[3]: content store: Device or resource busy
[1452816351.649735] kvs.err[3]: content_store_completion: Device or resource busy
```

garlick commented:
Did it empty out eventually?  (does content show 0 dirty?)

grondo commented:
It might be awhile:
```
$ while sleep 1; do flux getattr content-acct-dirty; done
118106
117870
117632
117395
117153
116913
116668
116421
116174
115923
115674
115420
115167
114908
114650
114390
114125
113861
113591
113322
113047
112773
112496
112213
111930
```

In this test, I should add, I'm running on a compute node so /tmp is tmpfs.

garlick commented:
I guess there's several issues interacting:
* the entire content cache backlog is sent to the backing store when it is loaded, one async rpc for each entry (many rpcs outstanding then)
* the backing store modules naively do a commit for each store operation so they are pretty slow!
* the long backlog of rpcs on the sender might be slowing down the matching of responses to rpc continuations
* matchtag pool exhaustion

So it seems like we could get some relief by 
* limiting the number of outstanding store requests
* on the backing stores, either temporally batching stores into commits and sending all the responses back when the commit completes, or changing the protocol so there is a separate commit message

grondo commented:
the flush appeared to get stuck at about 86K remaining dirty entries.

I then did a `flux content flush` and got the `EBUSY` errors and now we're proceeding with writing out dirty entries:

```
$ flux content flush
[1452817785.869784] broker.err[0]: content store: Device or resource busy
[1452817785.869789] broker.err[0]: content store: Device or resource busy
[1452817785.869796] broker.err[0]: content store: Device or resource busy
[1452817785.869802] broker.err[0]: content store: Device or resource busy
[1452817785.869807] broker.err[0]: content store: Device or resource busy
flux: content.flush: Device or resource busy
```

Actually, now my `flux getattr` is hanging....

grondo commented:
@garlick: those sound like reasonable solutions. If we launch with a default backing store (content-sqlite) perhaps it is highly unlikely that content-* modules would be swapped in and out like I was doing.

Do you think for now just limiting the number of outstanding backing store requests would be easy to implement? Perhaps using the idle watcher could help lazily batch up requests when there is dirty content, and avoid starving the broker for other services? (just throwing ideas out there)

It would be fine with me to make a basic fix for this PR here, with further refinements down the road so we can get the functionality merged now.

grondo commented:
Just FYI, the `flux getattr` command which was hung finally woke up, but looks like dirty objects are not being flushed (actually not sure if this is expected)

```
$ while sleep 1; do flux getattr content-acct-dirty; done
21127
21127
21127
21127
21127
21127

```

garlick commented:
Rebased on current master and pushed a fix for the problem @grondo exposed, testing like this:
```
$ flux start -s4
$ for i in `seq 1 1000`; do flux wreckrun -n4 hostname; done
$ flux module load content-sqlite
$ watch flux lsattr -v
```
This shows the rank 0 cache stabilizing within about 3s, e.g. no dirty entries, and purged below thresholds.  I think this is probably not a bad interim fix.

Now back to memory leaks.

garlick commented:
OK, found the leak - all three were due to the same one line `json_object_put()` omission when the KVS cache was refactored.  (Thanks for the valgrind assist guys)

I've rebased with that fix and also a minor off by one error in the "flush batch limit" patch.

Maybe this is close to ready?

grondo commented:
Nice work! Yeah, probably ready in my book. 
As a quick aside, is there somewhere we're documenting the broker attrs?

garlick commented:
There is a list in flux_attr_get(3) but that seems like the wrong place.

Should it be in flux-broker(1) or flux-getattr(1)?

grondo commented:
Good questions :-)
In flux-getattr seems like the right place, unless the attrs are extensible and then I'm not sure where is appropriate.

garlick commented:
They are really associated with the broker.  I'll go ahead and move those attrs to flux-getattr(1) and add the new ones.  We can always move them again if we need to.

grondo commented:
Ok, I'm also fine with them where they are... could they be kept in a separate place and linked to both manpages?

SteVwonder commented:
@grondo, flux /path/to/valgrind

garlick commented:
Just pushed an update that documents broker attributes in a new man page, flux-broker-attributes(7), and added a section 7 to docs.   That seemed better than replicating a list in the various man pages.  What do you think @grondo?

grondo commented:
Sounds good to me!
If this is ready I'll take a peek in the morning and merge

garlick commented:
Let me rebase on @dongahns commit and then I guess it's ready.  Thanks.

trws commented:
Once this goes in, what will the default be?  Are we in a position to have the content-sqlite used by default?

garlick commented:
I was on the fence about having one of those modules loaded by default.  I can make it the default if people feel that's the way to go.  Or we could wait until the next round of changes to the KVS.  Thoughts?

grondo commented:
Would the point be moot if we have an idea of how to deal with #508?

grondo commented:
I'm inclined to pull this in now, and we can delay decision on changing default content-* module, if that is ok?

trws commented:
Yeah, fine by me to delay it.  Having #508 would certainly make it easier to deal with an alternate default and let users configure it with less pain.  I'm not sure what the default should be here either, though having it be something resilient to oom seems preferable at first glance.
